src/jit/hwintrinsic.cpp

```diff
@@ -138,4 +173,557 @@ CORINFO_CLASS_HANDLE Compiler::gtGetStructHandleForHWSIMD(var_types simdType, va
+    }
+    else
+    {
+        assert(HWIntrinsicInfo::BaseTypeFromSecondArg(id));
+        CORINFO_ARG_LIST_HANDLE secondArg = comp->info.compCompHnd->getArgNext(sig->args);
+        typeHnd                           = comp->info.compCompHnd->getArgClass(sig, secondArg);
+    }
+
+    unsigned  simdSize = 0;
+    var_types baseType = comp->getBaseTypeAndSizeOfSIMDType(typeHnd, &simdSize);
+    assert((simdSize > 0) && (baseType != TYP_UNKNOWN));
+    return simdSize;
+}
+
+//------------------------------------------------------------------------
+// lookupNumArgs: Gets the number of args for a given HWIntrinsic node
+//
+// Arguments:
+//    node -- The HWIntrinsic node to get the number of args for
+//
+// Return Value:
+//    The number of args for the HWIntrinsic associated with node
+int HWIntrinsicInfo::lookupNumArgs(const GenTreeHWIntrinsic* node)
+{
+    assert(node != nullptr);
+
+    NamedIntrinsic id      = node->gtHWIntrinsicId;
+    int            numArgs = lookupNumArgs(id);
+
+    if (numArgs >= 0)
+    {
+        return numArgs;
+    }
+
+    assert(numArgs == -1);
+
+    GenTree* op1 = node->gtGetOp1();
+
+    if (op1 == nullptr)
+    {
+        return 0;
+    }
+
+    if (op1->OperIsList())
+    {
+        GenTreeArgList* list = op1->AsArgList();
+        numArgs              = 0;
+
+        do
+        {
+            numArgs++;
+            list = list->Rest();
+        } while (list != nullptr);
+
+        return numArgs;
+    }
+
+    GenTree* op2 = node->gtGetOp2();
+
+    return (op2 == nullptr) ? 1 : 2;
+}
+
+
+//------------------------------------------------------------------------
+// lookupLastOp: Gets the last operand for a given HWIntrinsic node
+//
+// Arguments:
+//    node   -- The HWIntrinsic node to get the last operand for
+//
+// Return Value:
+//     The last operand for node
+GenTree* HWIntrinsicInfo::lookupLastOp(const GenTreeHWIntrinsic* node)
+{
+    assert(node != nullptr);
+
+    NamedIntrinsic id  = node->gtHWIntrinsicId;
+    GenTree*       op1 = node->gtGetOp1();
+
+    if (op1 == nullptr)
+    {
+        return nullptr;
+    }
+
+    if (op1->OperIsList())
+    {
+        GenTreeArgList* list = op1->AsArgList();
+        GenTree*        last;
+
+        do
+        {
+            last = list->Current();
+            list = list->Rest();
+        } while (list != nullptr);
+
+        return last;
+    }
+
+    GenTree* op2 = node->gtGetOp2();
+
+    return (op2 == nullptr) ? op1 : op2;
+}
+
+//------------------------------------------------------------------------
+// isImmOp: Gets a value that indicates whether the HWIntrinsic node has an imm operand
+//
+// Arguments:
+//    id -- The NamedIntrinsic associated with the HWIntrinsic to lookup
+//    op -- The operand to check
+//
+// Return Value:
+//     true if the node has an imm operand; otherwise, false
+bool HWIntrinsicInfo::isImmOp(NamedIntrinsic id, const GenTree* op)
```

Comment:
Some of these methods aren't used by ARM anywhere yet, but they should still be applicable in a few spots.